src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,763 @@
+
+static const int64_t AUTOADAPT_INTERVAL_MS = 100;
+
+static bool is_excluded(JavaThread* thread) {
+  return thread->is_hidden_from_external_view() || thread->jfr_thread_local()->is_excluded();
+}
+
+static JavaThread* get_java_thread_if_valid() {
+  Thread* raw_thread = Thread::current_or_null_safe();
+  JavaThread* jt;
+  if (raw_thread == nullptr || !raw_thread->is_Java_thread()) { // this can happen due to the high level of parralelism
+    return nullptr;
+  }
+  if ((jt = JavaThread::cast(raw_thread))->is_exiting()) {
+    return nullptr;
+  }
+
+  if (is_excluded(jt)) {
+    return nullptr;
+  }
+  return jt;
+}
+
+JfrCPUTimeTraceQueue::JfrCPUTimeTraceQueue(u4 capacity) :
+  _capacity(capacity), _head(0), _lost_samples(0) {
+  _data = JfrCHeapObj::new_array<JfrCPUTimeSampleRequest>(capacity);
+}
+
+JfrCPUTimeTraceQueue::~JfrCPUTimeTraceQueue() {
+  JfrCHeapObj::free(_data, _capacity * sizeof(JfrCPUTimeSampleRequest));
+}
+
+bool JfrCPUTimeTraceQueue::enqueue(JfrCPUTimeSampleRequest& request) {
+  assert(JavaThread::current()->jfr_thread_local()->is_cpu_time_jfr_enqueue_locked(), "invariant");
+  u4 elementIndex;
+  do {
+    elementIndex = Atomic::load_acquire(&_head);
+    if (elementIndex >= _capacity) {
+      return false;
+    }
+  } while (Atomic::cmpxchg(&_head, elementIndex, elementIndex + 1) != elementIndex);
+  _data[elementIndex] = request;
+  return true;
+}
+
+JfrCPUTimeSampleRequest& JfrCPUTimeTraceQueue::at(u4 index) {
+  assert(index < _head, "invariant");
+  return _data[index];
+}
+
+volatile u4 _lost_samples_sum = 0;
+
+u4 JfrCPUTimeTraceQueue::size() const {
+  return Atomic::load(&_head);
+}
+
+void JfrCPUTimeTraceQueue::set_size(u4 size) {
+  Atomic::release_store(&_head, size);
+}
+
+u4 JfrCPUTimeTraceQueue::capacity() const {
+  return _capacity;
+}
+
+void JfrCPUTimeTraceQueue::set_capacity(u4 capacity) {
+  _head = 0;
+  JfrCPUTimeSampleRequest* new_data = JfrCHeapObj::new_array<JfrCPUTimeSampleRequest>(capacity);
+  JfrCHeapObj::free(_data, _capacity * sizeof(JfrCPUTimeSampleRequest));
+  _data = new_data;
+  _capacity = capacity;
+}
+
+bool JfrCPUTimeTraceQueue::is_full() const {
+  return Atomic::load_acquire(&_head) >= _capacity;
+}
+
+bool JfrCPUTimeTraceQueue::is_empty() const {
+  return Atomic::load_acquire(&_head) == 0;
+}
+
+s4 JfrCPUTimeTraceQueue::lost_samples() const {
+  return Atomic::load_acquire(&_lost_samples);
+}
+
+void JfrCPUTimeTraceQueue::increment_lost_samples() {
+  Atomic::inc(&_lost_samples_sum);
+  Atomic::inc(&_lost_samples);
+}
+
+u4 JfrCPUTimeTraceQueue::get_and_reset_lost_samples() {
+  s4 lost_samples = Atomic::load_acquire(&_lost_samples);
+  s4 new_lost_samples;
+  while ((new_lost_samples = Atomic::cmpxchg(&_lost_samples, lost_samples, 0)) != lost_samples) {
+    lost_samples = new_lost_samples;
+  }
+  return lost_samples;
+}
+
+void JfrCPUTimeTraceQueue::ensure_capacity(u4 capacity) {
+  if (capacity != _capacity) {
+    set_capacity(capacity);
+  }
+}
```

Comment:
Seems an odd definition - typically `ensure_capacity` will grow a data structure to ensure it has sufficient capacity, and if already larger than needed that is fine. Suggestion `change_capacity`, or more traditionally `resize`?